File: python/zebra_puzzle.py
from itertools import permutations

colors = ["red", "green", "ivory", "yellow", "blue"]
drinks = ["tea", "milk", "water", "juice", "coffee"]
pets = ["dog", "snails", "fox", "zebra", "horse"]
smokes = ["Old Gold", "Kools", "Chesterfields", "Lucky Strike", "Parliaments"]
nationalities = ["Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese"]

solution = next(
    list(zip(color, drink, pet, smoke, nationality))
    for color in permutations(colors)
    # 6
    if color.index("green") - color.index("ivory") == 1
    for nationality in permutations(nationalities)
    # 2
    if color.index("red") == nationality.index("Englishman")
    # 15
    if abs(color.index("blue") - nationality.index("Norwegian")) == 1
    for drink in permutations(drinks)
    # 9
    if drink.index("milk") == 2
    # 4
    if drink.index("coffee") == color.index("green")
    # 10
    if nationality.index("Norwegian") == 0
    # 5
    if nationality.index("Ukrainian") == drink.index("tea")
    for pet in permutations(pets)
    # 3
    if pet.index("dog") == nationality.index("Spaniard")
    for smoke in permutations(smokes)
    # 7
    if smoke.index("Old Gold") == pet.index("snails")
    # 8
    if smoke.index("Kools") == color.index("yellow")
    # 11
    if abs(pet.index("fox") - smoke.index("Chesterfields")) == 1
    # 12
    if abs(pet.index("horse") - smoke.index("Kools")) == 1
    # 13
    if smoke.index("Lucky Strike") == drink.index("juice")
    # 14
    if nationality.index("Japanese") == smoke.index("Parliaments")
)


def drinks_water():
    return [house for house in solution if house[1] == "water"][0][-1]


def owns_zebra():
    return [house for house in solution if house[2] == "zebra"][0][-1]

File: elixir/zebra_puzzle.ex
defmodule ZebraPuzzle do
  @nations [:englishman, :ukrainian, :spaniard, :norwegian, :japanese]
  @colors [:red, :green, :ivory, :yellow, :blue]
  @pets [:dog, :snails, :fox, :horse, :zebra]
  @drinks [:coffee, :tea, :milk, :juice, :water]
  @smokes [:old_gold, :kools, :chesterfields, :lucky_strike, :parliaments]

  @doc """
  Determine who drinks the water
  """
  @spec drinks_water() :: atom
  def drinks_water() do
    [nations, _, _, drinks, _] = solution()

    Enum.zip(nations, drinks)
    |> Enum.find(&match?({_, :water}, &1))
    |> elem(0)
  end

  @doc """
  Determine who owns the zebra
  """
  @spec owns_zebra() :: atom
  def owns_zebra() do
    [nations, _, pets, _, _] = solution()

    Enum.zip(nations, pets)
    |> Enum.find(&match?({_, :zebra}, &1))
    |> elem(0)
  end

  defp solution() do
    for nations <- permutations(@nations),
        match?([:norwegian | _], nations),
        colors <- permutations(@colors),
        right?(colors, :green, colors, :ivory),
        equal?(colors, :red, nations, :englishman),
        next_to?(nations, :norwegian, colors, :blue),
        pets <- permutations(@pets),
        equal?(nations, :spaniard, pets, :dog),
        drinks <- permutations(@drinks),
        match?([_, _, :milk, _, _], drinks),
        equal?(nations, :ukrainian, drinks, :tea),
        equal?(drinks, :coffee, colors, :green),
        smokes <- permutations(@smokes),
        next_to?(smokes, :chesterfields, pets, :fox),
        next_to?(smokes, :kools, pets, :horse),
        equal?(smokes, :lucky_strike, drinks, :juice),
        equal?(smokes, :kools, colors, :yellow),
        equal?(nations, :japanese, smokes, :parliaments),
        equal?(smokes, :old_gold, pets, :snails) do
      [nations, colors, pets, drinks, smokes]
    end
    |> hd()
  end

  defp equal?(xs, x, ys, y) do
    check(xs, x, ys, y, &==/2)
  end

  defp next_to?(xs, x, ys, y) do
    check(xs, x, ys, y, &(abs(&1 - &2) == 1))
  end

  defp right?(xs, x, ys, y) do
    check(xs, x, ys, y, &(&1 + 1 == &2))
  end

  defp check(xs, x, ys, y, f) do
    i = Enum.find_index(xs, &(&1 == x))
    j = Enum.find_index(ys, &(&1 == y))
    f.(i, j)
  end

  defp permutations(enumerable) do
    enumerable
    |> to_map()
    |> make_permutations()
  end

  defp to_map(enum), do: to_map(enum, fn _item, item_index -> item_index end)

  defp to_map(enum, key_function) when is_function(key_function, 1) do
    enum
    |> Stream.map(fn item -> {key_function.(item), item} end)
    |> Enum.into(%{})
  end

  defp to_map(enum, key_function) when is_function(key_function, 2) do
    enum
    |> Stream.with_index()
    |> Stream.map(fn {item, item_index} -> {key_function.(item, item_index), item} end)
    |> Enum.into(%{})
  end

  defp make_permutations(map) when map_size(map) == 0 do
    [[]]
  end

  defp make_permutations(map) when is_map(map) do
    map
    |> Stream.flat_map(fn {index, next} ->
      Map.delete(map, index)
      |> make_permutations()
      |> Stream.map(&[next | &1])
    end)
  end
end
